src/NServiceBus.Core/Receiving/ReceiveComponent.cs

```diff
@@ -211,7 +220,67 @@ IPushMessages BuildMessagePump()
             return messagePumpFactory();
         }
 
-        ReceiveConfiguration configuration;
+        static void LoadMessageHandlers(Configuration configuration, List<Type> orderedTypes, IConfigureComponents container, IList<Type> availableTypes)
+        {
+            var types = new List<Type>(availableTypes);
+
+            foreach (var t in orderedTypes)
+            {
+                types.Remove(t);
+            }
+
+            types.InsertRange(0, orderedTypes);
+
+            ConfigureMessageHandlersIn(configuration, types, container);
+        }
+
+        static void ConfigureMessageHandlersIn(Configuration configuration, IEnumerable<Type> types, IConfigureComponents container)
+        {
+            var handlerRegistry = configuration.MessageHandlerRegistry ?? new MessageHandlerRegistry();
+
+            foreach (var t in types.Where(IsMessageHandler))
+            {
+                container.ConfigureComponent(t, DependencyLifecycle.InstancePerUnitOfWork);
+                handlerRegistry.RegisterHandler(t);
+            }
+
+            container.RegisterSingleton(handlerRegistry);
+        }
+
+        public static bool IsMessageHandler(Type type)
+        {
+            if (type.IsAbstract || type.IsGenericTypeDefinition)
+            {
+                return false;
+            }
+
+            return type.GetInterfaces()
+                .Where(@interface => @interface.IsGenericType)
+                .Select(@interface => @interface.GetGenericTypeDefinition())
+                .Any(genericTypeDef => genericTypeDef == IHandleMessagesType);
+        }
+
+        public class Configuration
+        {
+            public Configuration(SettingsHolder settings)
+            {
+                this.settings = settings;
+            }
+
+            public List<Type> ExecuteTheseHandlersFirst
```

Comment:
should we make the getter `GetOrCreate` and remove the setter?

Reply:
I don't mind. When reviewed with @andreasohlund we decided to keep it as the original. Hence the `GetOrDefault()`.

Reply:
Changed.